File: apps/android/protocol/src/main/kotlin/dev/shruti/protocol/Protocol.kt
package dev.shruti.protocol

/**
 * On-wire packet header for phone -> laptop audio frames.
 *
 * Wire format (30 bytes):
 *   [magic: u32][version: u8][type: u8][flags: u8][phone_id: u8]
 *   [sequence: u32][sample_rate: u32][sample_count: u16][reserved: u32]
 *   [timestamp_us: u64]
 *
 * The Python reference implementation lives in
 * apps/laptop/shruti_array/protocol.py and must stay byte-compatible.
 *
 * Magic:   0x53555254 ('SHRT' little-endian)
 * Version: 1
 * Type:    0x01 AUDIO_FRAME | 0x02 CHIRP_ECHO | 0x03 HEARTBEAT
 * Flags:   bit 0 DROPPED | bit 1 LAST
 */
object Protocol {
    const val MAGIC: Int = 0x53555254
    const val VERSION: Int = 1
    const val HEADER_SIZE: Int = 30
    const val CRC_SIZE: Int = 4
    const val MAX_PAYLOAD_SAMPLES: Int = 16_384

    const val TYPE_AUDIO_FRAME: Int = 0x01
    const val TYPE_CHIRP_ECHO: Int = 0x02
    const val TYPE_HEARTBEAT: Int = 0x03

    const val FLAG_DROPPED: Int = 1 shl 0
    const val FLAG_LAST: Int = 1 shl 1

    // --- CRC-32C (Castagnoli, polynomial 0x1EDC6F41) -------------------------
    // Table-driven, bit-reflected. Verified at object init against
    // 0xE3069283 for the input "123456789" (RFC 7143 / iSCSI test vector).
    private const val CRC32C_POLY: Long = 0x82F63B78L

    private val CRC32C_TABLE: IntArray = buildCrc32cTable()

    private fun buildCrc32cTable(): IntArray {
        val table = IntArray(256)
        for (i in 0 until 256) {
            var crc = i
            repeat(8) {
                crc = if (crc and 1 != 0) (crc ushr 1) xor CRC32C_POLY.toInt()
                else crc ushr 1
            }
            table[i] = crc
        }
        return table
    }

    @JvmStatic
    fun crc32c(data: ByteArray, init: Int = 0xFFFFFFFF.toInt()): Int {
        var crc = init
        for (b in data) {
            val idx = (crc xor b.toInt()) and 0xFF
            crc = (crc ushr 8) xor CRC32C_TABLE[idx]
        }
        return crc xor 0xFFFFFFFF.toInt()
    }

    init {
        val selftest = crc32c("123456789".toByteArray(Charsets.US_ASCII))
        require(selftest == 0xE3069283.toInt()) {
            "CRC-32C self-test failed: got 0x${selftest.toUInt().toString(16)}, expected 0xe3069283"
        }
    }
}

class ProtocolError(message: String) : RuntimeException(message)

data class Header(
    val phoneId: Int,
    val sequence: Int,
    val sampleRateHz: Int,
    val sampleCount: Int,
    val timestampUs: Long,
    val packetType: Int,
    val flags: Int,
) {
    fun pack(): ByteArray {
        val buf = ByteArray(Protocol.HEADER_SIZE)
        val dv = java.nio.ByteBuffer.wrap(buf).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        dv.putInt(Protocol.MAGIC)
        dv.put(Protocol.VERSION.toByte())
        dv.put(packetType.toByte())
        dv.put((flags and 0xFF).toByte())
        dv.put((phoneId and 0xFF).toByte())
        dv.putInt(sequence)
        dv.putInt(sampleRateHz)
        dv.putShort(sampleCount.toShort())
        dv.putInt(0) // reserved
        dv.putLong(timestampUs)
        return buf
    }
}

private val HEADER_STRUCT = java.nio.ByteBuffer
    .allocate(Protocol.HEADER_SIZE)
    .order(java.nio.ByteOrder.LITTLE_ENDIAN)

fun parseHeader(packet: ByteArray, offset: Int = 0): Header {
    if (packet.size - offset < Protocol.HEADER_SIZE) {
        throw ProtocolError("buffer shorter than header")
    }
    val dv = java.nio.ByteBuffer.wrap(packet, offset, Protocol.HEADER_SIZE)
        .order(java.nio.ByteOrder.LITTLE_ENDIAN)
    val magic = dv.int
    val version = dv.get().toInt() and 0xFF
    val ptype = dv.get().toInt() and 0xFF
    val flags = dv.get().toInt() and 0xFF
    val phoneId = dv.get().toInt() and 0xFF
    val seq = dv.int
    val sr = dv.int
    val sc = dv.short.toInt() and 0xFFFF
    dv.int // reserved
    val ts = dv.long
    if (magic != Protocol.MAGIC) throw ProtocolError("bad magic 0x${magic.toString(16)}")
    if (version != Protocol.VERSION) throw ProtocolError("unsupported version $version")
    if (ptype !in setOf(Protocol.TYPE_AUDIO_FRAME, Protocol.TYPE_CHIRP_ECHO, Protocol.TYPE_HEARTBEAT)) {
        throw ProtocolError("unknown packet type $ptype")
    }
    if (sc > Protocol.MAX_PAYLOAD_SAMPLES) {
        throw ProtocolError("sample_count $sc exceeds maximum")
    }
    return Header(phoneId, seq, sr, sc, ts, ptype, flags)
}

/** Build a complete packet (header + payload + CRC) from PCM int16 samples. */
fun framePacket(
    phoneId: Int,
    sequence: Int,
    sampleRateHz: Int,
    samples: ByteArray, // little-endian int16 PCM, sample_count * 2 bytes
    timestampUs: Long,
    packetType: Int = Protocol.TYPE_AUDIO_FRAME,
    flags: Int = 0,
): ByteArray {
    if (samples.size % 2 != 0) throw ProtocolError("PCM payload must be int16-aligned")
    val sampleCount = samples.size / 2
    if (sampleCount > Protocol.MAX_PAYLOAD_SAMPLES) throw ProtocolError("payload too large")
    val header = Header(
        phoneId = phoneId,
        sequence = sequence,
        sampleRateHz = sampleRateHz,
        sampleCount = sampleCount,
        timestampUs = timestampUs,
        packetType = packetType,
        flags = flags,
    )
    val headerBytes = header.pack()
    val body = headerBytes + samples
    val crc = crc32c(body)
    val out = ByteArray(body.size + Protocol.CRC_SIZE)
    System.arraycopy(body, 0, out, 0, body.size)
    val crcBuf = java.nio.ByteBuffer.wrap(out, body.size, Protocol.CRC_SIZE)
        .order(java.nio.ByteOrder.LITTLE_ENDIAN)
    crcBuf.putInt(crc)
    return out
}

fun verifyPacket(packet: ByteArray): Header {
    if (packet.size < Protocol.HEADER_SIZE + Protocol.CRC_SIZE) {
        throw ProtocolError("packet shorter than header+CRC")
    }
    val header = parseHeader(packet)
    val payloadLen = header.sampleCount * 2
    val expected = Protocol.HEADER_SIZE + payloadLen + Protocol.CRC_SIZE
    if (packet.size != expected) {
        throw ProtocolError(
            "packet length ${packet.size} != expected $expected for ${header.sampleCount} samples"
        )
    }
    val dv = java.nio.ByteBuffer.wrap(
        packet, Protocol.HEADER_SIZE + payloadLen, Protocol.CRC_SIZE,
    ).order(java.nio.ByteOrder.LITTLE_ENDIAN)
    val actualCrc = dv.int
    val expectedCrc = crc32c(packet.copyOfRange(0, Protocol.HEADER_SIZE + payloadLen))
    if (actualCrc != expectedCrc) throw ProtocolError("CRC mismatch")
    return header
}

fun payload(packet: ByteArray): ByteArray {
    val header = verifyPacket(packet)
    return packet.copyOfRange(
        Protocol.HEADER_SIZE,
        Protocol.HEADER_SIZE + header.sampleCount * 2,
    )
}

// --- CRC-32C top-level convenience ----------------------------------------
// The implementation lives inside `object Protocol` so the self-test
// runs at class init; this thin wrapper preserves the existing
// top-level call sites (and the public API).
fun crc32c(data: ByteArray, init: Int = 0xFFFFFFFF.toInt()): Int =
    Protocol.crc32c(data, init)

File: apps/laptop/shruti_array/protocol.py
"""Shared packet protocol between Android sender and laptop receiver.

The protocol is deliberately tiny and binary: the array's whole real-time
budget is in the per-frame latency, so headers must be cheap to parse and
zero-copy where possible.

Wire format (single direction: phone -> laptop):

    [magic: u32][version: u8][type: u8][flags: u8][phone_id: u8]
    [sequence: u32][sample_rate: u32][sample_count: u16][reserved: u32]
    [timestamp_us: u64]
    [payload:     i16 * sample_count]      (little-endian PCM)
    [crc32c:      u32]                     (over header + payload)

Total header: 30 bytes (struct naturally packs; alignment of Q after H+I
ends up at offset 22 with no padding needed).

Magic:   0x53555254 ('SHRT' little-endian)
Version: 1
Type:
    0x01 AUDIO_FRAME   - payload is mono PCM at sample_rate
    0x02 CHIRP_ECHO    - payload is a chirp recording (alignment signal)
    0x03 HEARTBEAT     - payload is empty, timestamp_us is the wall clock
Flags:
    bit 0: dropped_frames  (set by sender when this frame skips samples)
    bit 1: last_frame      (set when sender is closing cleanly)

The on-wire size for a 20 ms frame at 48 kHz is:
    30 byte header + 960 * 2 byte payload + 4 byte crc = 1954 bytes
"""
from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Final

MAGIC: Final[int] = 0x53555254  # 'SHRT'
VERSION: Final[int] = 1
HEADER_SIZE: Final[int] = 30
CRC_SIZE: Final[int] = 4


class PacketType(IntEnum):
    AUDIO_FRAME = 0x01
    CHIRP_ECHO = 0x02
    HEARTBEAT = 0x03


class Flag(IntEnum):
    DROPPED = 1 << 0
    LAST = 1 << 1


_HEADER_STRUCT = struct.Struct("<IBBBBII H I Q")  # 30 bytes
if _HEADER_STRUCT.size != HEADER_SIZE:
    raise RuntimeError(
        f"Header struct size {_HEADER_STRUCT.size} != HEADER_SIZE {HEADER_SIZE}"
    )

# Maximum payload: 100 ms at 96 kHz mono = 9600 samples.
# We never produce that, but bounding the parser protects us from hostile input.
MAX_PAYLOAD_SAMPLES: Final[int] = 16_384


class ProtocolError(ValueError):
    """Raised on any malformed packet."""


@dataclass(frozen=True)
class Header:
    phone_id: int
    sequence: int
    sample_rate_hz: int
    sample_count: int
    timestamp_us: int
    packet_type: PacketType
    flags: int

    def pack(self) -> bytes:
        return _HEADER_STRUCT.pack(
            MAGIC,
            VERSION,
            int(self.packet_type),
            self.flags & 0xFF,
            self.phone_id & 0xFF,
            self.sequence,
            self.sample_rate_hz,
            self.sample_count,
            0,
            self.timestamp_us,
        )


def parse_header(buf: bytes, offset: int = 0) -> Header:
    if len(buf) - offset < HEADER_SIZE:
        raise ProtocolError("buffer shorter than header")
    magic, version, ptype, flags, phone_id, seq, sr, sc, _reserved, ts = (
        _HEADER_STRUCT.unpack_from(buf, offset)
    )
    if magic != MAGIC:
        raise ProtocolError(f"bad magic 0x{magic:08x}")
    if version != VERSION:
        raise ProtocolError(f"unsupported version {version}")
    if ptype not in PacketType._value2member_map_:
        raise ProtocolError(f"unknown packet type {ptype}")
    if sc > MAX_PAYLOAD_SAMPLES:
        raise ProtocolError(f"sample_count {sc} exceeds maximum")
    return Header(
        phone_id=phone_id,
        sequence=seq,
        sample_rate_hz=sr,
        sample_count=sc,
        timestamp_us=ts,
        packet_type=PacketType(ptype),
        flags=flags,
    )


def frame_packet(
    *,
    phone_id: int,
    sequence: int,
    sample_rate_hz: int,
    samples: bytes,
    timestamp_us: int,
    packet_type: PacketType = PacketType.AUDIO_FRAME,
    flags: int = 0,
) -> bytes:
    """Build a complete packet (header + payload + CRC) from a PCM payload.

    `samples` is raw little-endian int16 PCM, exactly `sample_count * 2` bytes.
    The CRC is computed over header + payload using CRC-32C (Castagnoli),
    which is also the polynomial used by iSCSI and many NIC offloads.
    """
    if len(samples) % 2 != 0:
        raise ProtocolError("PCM payload must be int16-aligned")
    sample_count = len(samples) // 2
    if sample_count > MAX_PAYLOAD_SAMPLES:
        raise ProtocolError("payload too large")
    header = Header(
        phone_id=phone_id,
        sequence=sequence,
        sample_rate_hz=sample_rate_hz,
        sample_count=sample_count,
        timestamp_us=timestamp_us,
        packet_type=packet_type,
        flags=flags,
    )
    header_bytes = header.pack()
    crc = _crc32c(header_bytes + samples)
    return header_bytes + samples + struct.pack("<I", crc)


def verify_packet(packet: bytes) -> Header:
    """Parse + CRC-verify a complete packet. Returns the header."""
    if len(packet) < HEADER_SIZE + CRC_SIZE:
        raise ProtocolError("packet shorter than header+CRC")
    header = parse_header(packet)
    payload_len = header.sample_count * 2
    expected = HEADER_SIZE + payload_len + CRC_SIZE
    if len(packet) != expected:
        raise ProtocolError(
            f"packet length {len(packet)} != expected {expected} for "
            f"{header.sample_count} samples"
        )
    actual_crc = struct.unpack_from("<I", packet, HEADER_SIZE + payload_len)[0]
    expected_crc = _crc32c(packet[: HEADER_SIZE + payload_len])
    if actual_crc != expected_crc:
        raise ProtocolError("CRC mismatch")
    return header


def payload(packet: bytes) -> bytes:
    """Return just the PCM payload bytes of a verified packet."""
    header = verify_packet(packet)
    return packet[HEADER_SIZE : HEADER_SIZE + header.sample_count * 2]


# --- CRC-32C (Castagnoli, polynomial 0x1EDC6F41) -----------------------------
# Table-driven, bit-reflected for the LSB-first algorithm. Verified at import
# time against the well-known check value 0xE3069283 for b"123456789"
# (the same vector used by the iSCSI spec, RFC 7143).
_CRC32C_POLY = 0x82F63B78  # bit-reflected Castagnoli


def _build_crc32c_table() -> tuple[int, ...]:
    table = []
    for i in range(256):
        crc = i
        for _ in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ _CRC32C_POLY
            else:
                crc >>= 1
        table.append(crc)
    return tuple(table)


_CRC32C_TABLE = _build_crc32c_table()


def _crc32c(data: bytes, init: int = 0xFFFFFFFF) -> int:
    crc = init
    for b in data:
        crc = (crc >> 8) ^ _CRC32C_TABLE[(crc ^ b) & 0xFF]
    return crc ^ 0xFFFFFFFF


# Self-test on import. The CRC is the only thing the protocol depends on for
# packet integrity; if it regresses, every packet becomes a coin flip.
_CRC32C_SELFTEST = _crc32c(b"123456789")
if _CRC32C_SELFTEST != 0xE3069283:
    raise RuntimeError(
        f"CRC-32C self-test failed: got 0x{_CRC32C_SELFTEST:08X}, "
        f"expected 0xE3069283. The Castagnoli implementation is broken."
    )
